### ISSUE TITLE
Compilation Stuck on Attributable.cpp.o

### ISSUE THREAD

**Issue by DrakeWhu** [ECP-WarpX/WarpX#4308]
Hi, first of all my name is Juan, I've been working in laser-plasma simulations for over a year with EPOCH and this year I decided to use WarpX.

I've been trying to compile WarpX on a virtual machine with Arch Linux, given that soon the cluster in my institution will get an upgrade and Arch is the OS of choice. I followed the standard installation steps, including setting up WarpX with Python integration. However, during the compilation process using `make -j$(nproc)`, it seems to get stuck on Attributable.cpp.o. I've waited for over an hour, but there doesn't seem to be any progress.

Steps to Reproduce:

- Clone the WarpX repository.
- Set up a Python virtual environment and install the required Python packages.
- Create a build directory and navigate to it.
- Run `cmake .. -DWarpX_PYTHON=ON -DWarpX_DIMS=3`
- Run `make -j$(nproc)`.

Expected Behavior:
The compilation should proceed without getting stuck and complete successfully.

Actual Behavior:
The compilation gets stuck on `Attributable.cpp.o` and doesn't progress after waiting for over an hour.

Additional Information:

- OS: Arch Linux (running on a virtual machine)
- Python version: 3.x
- Only python3 and pip were initially installed on the system.
- VM Resources: 1GB RAM and 1 CPU core dedicated.

Any guidance or suggestions would be greatly appreciated. Also I will leave the compilation running just in case it's just super slow on a VM.

**Comment by RemiLehe** [comment]
Thanks for reporting this issue.

Could you try to compile without parallelism, i.e. `make` instead of `make -j $(nproc)`? 
(This is because compiling in parallel sometimes makes it hard to see where the compilation is actually stuck.)

Thanks!

**Comment by DrakeWhu** [comment]
I already did and it apparently got stuck in the same file. I'm gonna try to compile it again to see if for some reason now gets stuck in other place. I'll comment again in a few hours.

Thank you.

EDIT: Yes, I just ran `make` and it got stuck in Attributable.cpp.o again

**Comment by ax3l** [comment]
@DrakeWhu Thank you for the report and welcome to WarpX! :)

Can you please save the whole CMake configuration and also the compilation output to a text file and post it here?
I would like to see details on your compilers, linkers, etc. that are contained in these details to diagnose the problem further.

The problem you see is in https://github.com/openPMD/openPMD-api (just FYI), but we can continue to support you here in this ticket for simplicity, because we automatically download a copy of openPMD-api to build WarpX.

You can, as the very last resort, simply disable openPMD via `-DWarpX_OPENPMD=OFF` in the CMake configuration step. But I would not recommend going there, because there is a simple fix.

> VM Resources: 1GB RAM and 1 CPU core dedicated.

As a first guess, the problem is likely a resource exhaustion for your compiler.
My general guidance so far:

1. try to compile without `-j` (stay serial). Your VM provided only 1 CPU and more compile processes will oversubscribe your resources.
2. Can you increase the VM RAM resources further? In particular, I would increase the RAM to >4 GB, if that is possible. *This is likely your main issue.*
3. Can you increase the VM CPU resources, e.g., to 2 or 4 cores? That way, you could be able to use `-j2` or `-j4` later on, if you have enough RAM left.

For background info, the WarpX executables and intermediate libraries are, depending on build mode, 100s of MBs to a few GB. Your compiler and linker will need to copy them around, optimize their logic, etc. to assemble the final application and Python modules, thus you really want to have several GBs of RAM to compile the project.

How large is the host machine of your VM? Is there a hard limit why you picked such a small setup (smaller than a [Raspberry PI 4 (2019)](https://www.raspberrypi.com/products/raspberry-pi-4-model-b/specifications/) pocket PC :) )?

**Comment by DrakeWhu** [comment]
Hi @ax3l ,

Thank you for your prompt response. I've attached the output from CMake and the compilation process as requested. The make output has a keyboard interrupt because it gets hung up at that point. 

Regarding the VM resources, I initially allocated 1GB of RAM and 1 CPU core as a starting point. I can try increasing the RAM to >4 GB and the CPU cores to 2 or 4 and see if that helps with the compilation. 
[make_output.txt](https://github.com/ECP-WarpX/WarpX/files/12663259/make_output.txt)
[cmake_output.txt](https://github.com/ECP-WarpX/WarpX/files/12663260/cmake_output.txt)

My host machine has 16GB RAM and 6 cores. I chose the initial VM just as a test, so I have no problem on dedicating more resources.

I'll try adjusting the VM resources and recompiling. I'll keep you updated on the progress.

Best regards,

Juan

EDIT: It looks like increasing the VM resources worked, it's compiling now, I'll keep you updated.

**Comment by DrakeWhu** [comment]
Hi @ax3l and WarpX team,

Thank you for your previous guidance. I've successfully compiled WarpX after adjusting the VM resources. However, I'm still encountering an issue when trying to run a 3D PICMI script. The error message indicates that the 3D dimensionality hasn't been compiled, even though I've specified the -DWarpX_DIMS=3 flag during the CMake configuration. I will create a new issue about this specific problem, and if you feel like this one should be closed, I'll leave it to you.

Thanks for the help,
Juan
